agent/grpc-external/limiter/limiter.go

```diff
@@ -0,0 +1,245 @@
+//	   stream) when it is closed.
+func (l *SessionLimiter) BeginSession() (*Session, error) {
+	if !l.hasCapacity() {
+		return nil, ErrCapacityReached
+	}
+
+	l.mu.Lock()
+	defer l.mu.Unlock()
+	return l.createSessionLocked(), nil
+}
+
+// Note: hasCapacity is *best effort*. As we do not hold l.mu it's possible that:
+//
+//	- max has changed by the time we compare it to inFlight.
+//	- inFlight < max now, but increases before we create a new session.
+//
+// This is acceptable for our uses, especially because excess sessions will
+// eventually be drained.
+func (l *SessionLimiter) hasCapacity() bool {
+	max := atomic.LoadUint32(&l.max)
+	if max == Unlimited {
+		return true
+	}
+
+	cur := atomic.LoadUint32(&l.inFlight)
+	return max > cur
+}
+
+// Note: overCapacity is *best effort*. As we do not hold l.mu it's possible that:
+//
+//	- max has changed by the time we compare it to inFlight.
+//	- inFlight > max now, but decreases before we terminate a session.
+func (l *SessionLimiter) overCapacity() bool {
+	max := atomic.LoadUint32(&l.max)
+	if max == Unlimited {
+		return false
+	}
+
+	cur := atomic.LoadUint32(&l.inFlight)
+	return cur > max
+}
+
+func (l *SessionLimiter) terminateSession() {
+	l.mu.Lock()
+	defer l.mu.Unlock()
+
+	idx := rand.Intn(len(l.sessionIDs))
+	id := l.sessionIDs[idx]
+	l.sessions[id].terminate()
+	l.deleteSessionLocked(idx)
+}
+
+func (l *SessionLimiter) createSessionLocked() *Session {
+	session := &Session{
+		l:      l,
+		id:     l.maxSessionID,
+		termCh: make(chan struct{}),
+	}
+
+	l.maxSessionID++
+	l.sessionIDs = append(l.sessionIDs, session.id)
+	l.sessions[session.id] = session
+
+	atomic.AddUint32(&l.inFlight, 1)
+
+	return session
+}
+
+func (l *SessionLimiter) deleteSessionLocked(idx int) {
+	delete(l.sessions, l.sessionIDs[idx])
+
+	l.sessionIDs[idx] = l.sessionIDs[len(l.sessionIDs)-1]
+	l.sessionIDs = l.sessionIDs[:len(l.sessionIDs)-1]
+
+	atomic.AddUint32(&l.inFlight, ^uint32(0))
+}
+
+// Session allows its holder to perform an operation (e.g. serve a gRPC stream)
+// concurrenly with other session-holders. Sessions may be terminated abruptly
+// by the SessionLimiter, so it is the responsibility of the holder to receive
+// on the Terminated channel and halt the operation when it is closed.
+type Session struct {
+	l *SessionLimiter
+
+	id     uint64
+	termCh chan struct{}
+
+	// done ensures that if both End and terminate are called on a session, only
+	// the first call will have effect - this is important as both will attempt to
+	// clean up the session's state. It also ensures calling End more than once is
+	// a no-op.
+	done uint32
+}
+
+// End the session.
+//
+// This MUST be called when the session-holder is done (e.g. the gRPC stream
+// is closed).
+func (s *Session) End() {
+	if !atomic.CompareAndSwapUint32(&s.done, 0, 1) {
+		return
+	}
+
+	s.l.mu.Lock()
+	defer s.l.mu.Unlock()
+
+	idx := -1
+	for i, id := range s.l.sessionIDs {
+		if id == s.id {
+			idx = i
+			break
+		}
+	}
```

Comment:
This linear search of `sessionIDs` isn't great, but because we grow and shrink the slice as sessions are added and removed we can't track the index in the `Session` struct.

Given `n` is bounded by the number of open xDS streams, I don't think it could become large enough to become a problem worth solving with a more complicated data structure, but I'm open to feedback!

Reply:
Actually thinking about it, because session IDs increase monotonically, we could fairly easily store them sorted and do a binary search instead `O(log n)`.